test-utils/chain-spec-builder/src/main.rs

```diff
@@ -1,49 +1,263 @@
+			keystore_path.join(format!("auth-{}", n)),
+			None,
+		).map_err(|err| err.to_string())?;
+
+		let (_, _, grandpa, babe, im_online) =
+			chain_spec::get_authority_keys_from_seed(seed);
+
+		let insert_key = |key_type, public| {
+			keystore.write().insert_unknown(
+				key_type,
+				&format!("//{}", seed),
+				public,
+			).map_err(|_| format!("Failed to insert key: {}", grandpa))
+		};
+
+		insert_key(
+			primitives::crypto::key_types::BABE,
+			babe.as_slice(),
+		)?;
+
+		insert_key(
+			primitives::crypto::key_types::GRANDPA,
+			grandpa.as_slice(),
+		)?;
+
+		insert_key(
+			primitives::crypto::key_types::IM_ONLINE,
+			im_online.as_slice(),
+		)?;
+	}
+
+	Ok(())
+}
+
+fn print_seeds(
+	authority_seeds: &[String],
+	endowed_seeds: &[String],
+	sudo_seed: &str,
+) {
+	let header = Style::new().bold().underline();
+	let entry = Style::new().bold();
+
+	println!("{}", header.paint("Authority seeds"));
+
+	for (n, seed) in authority_seeds.iter().enumerate() {
+		println!("{} //{}",
+			entry.paint(format!("auth-{}:", n)),
+			seed,
+		);
+	}
+
+	println!();
+
+	if !endowed_seeds.is_empty() {
+		println!("{}", header.paint("Endowed seeds"));
+		for (n, seed) in endowed_seeds.iter().enumerate() {
+			println!("{} //{}",
+				entry.paint(format!("endowed-{}:", n)),
+				seed,
+			);
+		}
+
+		println!();
+	}
+
+	println!("{}", header.paint("Sudo seed"));
+	println!("//{}", sudo_seed);
 }
 
-fn main() {
-	let json = generate_chain_spec();
-	println!("{}", json);
+fn main() -> Result<(), String> {
+	let builder = ChainSpecBuilder::from_args();
+	let chain_spec_path = builder.chain_spec_path().to_path_buf();
+
+	let (authority_seeds, endowed_accounts, sudo_account) = match builder {
+		ChainSpecBuilder::Generate { authorities, endowed, keystore_path, .. } => {
+			let authorities = authorities.max(1);
+			let rand_str = || -> String {
+				OsRng.sample_iter(&Alphanumeric)
+					.take(32)
+					.collect()
+			};
+
+			let authority_seeds = (0..authorities).map(|_| rand_str()).collect::<Vec<_>>();
+			let endowed_seeds = (0..endowed).map(|_| rand_str()).collect::<Vec<_>>();
+			let sudo_seed = rand_str();
+
+			print_seeds(
+				&authority_seeds,
+				&endowed_seeds,
+				&sudo_seed,
+			);
+
+			if let Some(keystore_path) = keystore_path {
+				generate_authority_keys_and_store(
+					&authority_seeds,
+					&keystore_path,
+				)?;
+			}
+
+			let endowed_accounts = endowed_seeds.iter().map(|seed| {
+				chain_spec::get_from_seed::<AccountId>(seed)
+					.to_ss58check()
+			}).collect();
+
+			let sudo_account = chain_spec::get_from_seed::<AccountId>(&sudo_seed)
+				.to_ss58check();
+
+			(authority_seeds, endowed_accounts, sudo_account)
+		},
+		ChainSpecBuilder::New { authority_seeds, endowed_accounts, sudo_account, .. } => {
+			(authority_seeds, endowed_accounts, sudo_account)
+		},
+	};
+
+	let json = generate_chain_spec(
+		authority_seeds,
+		endowed_accounts,
+		sudo_account,
+	)?;
+
+	let mut file = File::create(chain_spec_path)
```

Comment:
These 6 lines can be abbreviated with https://doc.rust-lang.org/std/fs/fn.write.html

So
```
fs::write(chain_spec_path, json).map_err(ToString::to_string)
```